keps/sig-node/2570-memory-qos/README.md

```diff
@@ -0,0 +1,121 @@
+| -------- | -------- | -------- | -------- | -------- |
+| Guaranteed | y | y | requests | n |
+| Burstable | n | y | n | requests |
+
+### User Stories (Optional)
+- Workload is memory-sensitive, this feature can retain memory to reduce allocation latency
+- Workload is burstable, this feature can try best to protect request memory and decrease probability of memory being reclaimed
+
+### Comparison with Memory Manager
+The Memory Manager is a new component of kubelet ecosystem proposed to enable single-NUMA and multi-NUMA guaranteed memory allocation at topology level. And this Memory QoS proposal mainly uses the cgrups v2 mechanism to improve the quality of memory allocation, thereby improving the memory qos of guanranteed and burstable pods and even entire node. 
+See also https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1769-memory-manager
+
+### Risks and Mitigations
+- `Guaranteed` pod memory reservation could reduce node memory supply
+
+## Design Details
+### Feature Gate
+To set `--feature-gates=MemoryQOS=true` to enable the feature.
+
+### Rules
+1. if every containers in pod is set `requests.memory` and `limits.memory` (guaranteed pod), we set `memory.min = requests.memory` at container cgroup level and set `memory.min = sum(pod.spec.containers.resources.requests[memory])` at pod cgroup level
+2. if some containers in pod are set `requests.memory` or `limit.memory` (burstable pod), we set `memory.low = sum(pod.spec.containers.resources.requests[memory])` at pod cgroup level
+
+### Interactive
+New `Unified` field will be added in both CRI and QoS Manager for cgroups v2 extra parameters. It is recommended to keep same meaning with OCI runtime specs https://github.com/opencontainers/runtime-spec/pull/1040
+
+- container level for guaranteed pod: `Unified` added in `LinuxContainerResources`
+- pod level for burstable pod: `Unified` added in `cm.ResourceConfig` 
+
+### Workflow
+#### Guaranteed Pod
+![](./container-level-wf.png)
+#### Burstable Pod
+![](./pod-level-wf.png)
+
+### CGroups v2 Support
+After Kubernetes v1.19, kubelet can identify cgroups v2 and do the convention. Since [v1.0.0-rc93](https://github.com/opencontainers/runc/releases/tag/v1.0.0-rc93), runc supports `Unified` to pass through cgroups v2 parameters. So we use this variable to pass `memory.low / memory.min` when cgroups v2 mode is detected.
+
+### CGroup Hierarchy 
+Guaranteed pod:
+```
+/cgroup2/kubepods/pod<UID>/memory.min = sum(pod.spec.containers.resources.requests[memory])
+/cgroup2/kubepods/pod<UID>/<container-id>/memory.min = pod.spec.containers[i].resources.requests[memory]
+```
+
+Burstable pod:
+```
+/cgroup2/kubepods/burstable/pod<UID>/memory.low = sum(pod.spec.containers.resources.requests[memory])
+```
+
+### Container Runtime Interface (CRI) Changes
+Since we use `Unified` to place cgroups v2 extra parameters, CRI `LinuxContainerResources` should be added a new field for that.
```

Comment:
might be worth to note that Unified in CRI api is basically passthrough for OCI spec `Unified` field and has same semantics: https://github.com/opencontainers/runtime-spec/pull/1040